### PR TITLE
i18n

### LINKED ISSUE #288: i18n capabilities
Hi,

My friends and I would like to translate this resource into Ukrainian. I saw in `CONTRIBUTING.md` that you encourage contributing translations. But I don't see any localisation/internationalisation infrastructure in the repo.

Could you please share any plans on changing the folder structure (like having one folder per locale with the same set of files as English)? Or maybe you plan using any i18n `_` markers?

Looking forward to hearing your thoughts :)

Warm regards,
Sviatoslav.

### PR DESCRIPTION
**Pull request by bkeepers**
This is a work in progress to make it possible to translate the site. Any and all feedback on how to make this easier is welcome.

## Implementation

This site uses [Jekyll](http://jekyllrb.com/) and is hosted on [GitHub Pages](https://pages.github.com/). There is a Jekyll [plugin for multiple languages](https://github.com/Anthony-Gaudino/jekyll-multiple-languages-plugin), but it's not supported on GitHub Pages. So we'll have to manually add the facilities for supporting translations.

Currently, I'm thinking that the infrastructure for doing translations will be in this repository, and the translations themselves will exist as forks. Completed translations will be given a subdomain for their locale (like `fr.opensource.guide`) and be listed on the website. We'll make it as easy as possible for translations to get updates, but ultimately it will be the translation maintainer's responsibility to keep up to date.

Thoughts?

## TODO

- [x] Structure for localizing strings (`_data/locale/zz.yml`)
- [x] Update all strings in the markup
- [x] figure out structure for translating the guides themselves
- [x] ~~Tests for missing and unused localization strings~~
- [x] Figure out pattern for translating aria labels on octicons
- [x] Add locale switcher on the site (#354)
- [ ] Docs/process for translating the twitter cards
- [ ] Remove or make "Section x" headings localizable (#364)
- [ ] Complete one translation to make sure the infrastructure is sufficient (#299)
- [x] Document the translation process in ```CONTRIBUTING.md```

Fixes #288

### CODE REVIEW

**Comment by liadbiz** [comment]
@bkeepers 
I am working on translation of chinese version. I will post on my personal blog before i18n is supported on official website.
thanks for your work!

**Comment by webknjaz** [comment]
@bkeepers
I think that keeping `gh-pages` main branch makes it complicated to support different repo configs. Perhaps the branch with published web-site should be generated via CI and it should be possible to keep all the stuff related to original sources and sources of translated articles without interferring.
e.g. Having `.travis.yml` in sync and supporting multi-language `_config.yml`.

What do you think about such setup?

**Comment by bkeepers** [comment]
@webknjaz I'm not sure I understand what you're suggesting.

The problem is that Jekyll doesn't support i18n natively, and GitHub Pages doesn't support any of the plugins that try to add it. For now, I would like to avoid a custom build process, so I am aiming to make it as easy as possible for people to create forks of the website with translations, but in a way that will make it easy to merge back together later if/when jekyll and GitHub Pages gets native i18n support.

In my mind, the easiest way to do this is for each language to have a canonical fork (e.g; you would maintain the Ukranian fork on your own GitHub account) that takes over the `gh-pages` branch. Each translation will get an official subdomain (`uk.opensource.guide`). To the end-user, it will look like one website, even though each fork is technically a completely separate site on GitHub Pages.

Does that make sense? I'd love feedback on this approach.

**Comment by webknjaz** [comment]
@bkeepers oh, I was thinking the build process could rely on some env var or repo naming convention, so that it would just pick language depending on that condition. I usually do the build via Travis CI and push to gh-pages generating `CNAME` file dynamically.
Can this be an option? This would require just a few configuration changes from the fork owner (probably just enable Travis CI and and an env var).
This way the forks could be periodically merged back to upstream w/o any conflicts like having different `CNAME`/`_config.yml` etc.
On the other hand each time when upstream changes are being absorbed by forks this option will eliminate same issues.

**Comment by bkeepers** [comment]
> oh, I was thinking the build process could rely on some env var or repo naming convention, so that it would just pick language depending on that condition. I usually do the build via Travis CI and push to gh-pages generating `CNAME` file dynamically. Can this be an option?

I'd really like to avoid custom build stuff for now. Let's get the actual translations done first and then we can figure out how to optimize everything else.

**Comment by webknjaz** [comment]
Sure, there are 4 files in progress of translation already :)

**Comment by joelhandwell** [comment]
@bkeepers have you considered to utilize open source friendly translation management system like https://www.transifex.com/?

**Comment by webknjaz** [comment]
@joelhandwell AFAIK that tool is designed for interface localisation purposes (and isn't free) and doesn't fit article translation requirements

**Comment by bkeepers** [comment]
> @bkeepers have you considered to utilize open source friendly translation management system like https://www.transifex.com/?

@joelhandwell I've never used anything like that. Does it work well with vanilla Jekyll on GitHub Pages?

**Comment by joelhandwell** [comment]
> @webknjaz AFAIK that tool is designed for interface localisation purposes (and isn't free) and doesn't fit article translation requirements

I agree that Transifex is primary designed for interface localization. However as we can see [Creative Commons](https://creativecommons.org/licenses/by/4.0/) [using](https://wiki.creativecommons.org/wiki/Translating_CC_Deeds) we can say it's designed for not only for interface, but also for articles. [The Data Journalism Handbook](https://github.com/Mortimerp9/TheDataJournalismHandbook) also [chose it to translate](https://www.transifex.com/liliana.bounegru/the-data-journalism-handbook/) articles.

> (and isn't free)

According to their [FAQ](https://docs.transifex.com/faq/all) "What is considered an Open Source project?":

> Transifex offers an Open Source Program to Open Source projects that have no funding, revenue, and/or commercialization model. The intent is to support small and independent Open Source projects. If your project does not fall in this category, please [contact us](https://www.transifex.com/contact/) to discuss the appropriate pricing.
> 
> Open Source projects are ones with publicly available source codes licensed under an OSI-approved license, such as GPL or BSD.

So if this project is funded by Github by any means, appropriate pricing may be discussed, otherwise if they approve license of this project are ok, I guess it's free.

I do not want to restrict options to only in Transifex. There is alternatives for Transifex for example [Zanata](http://zanata.org/) with its [source ](https://github.com/zanata/zanata-server). Big project like OpenStack [migrated from Transifex to Zanata](http://princessleia.com/journal/2015/09/the-migration-of-openstack-translations-to-zanata/). For your info, here is [OpenStack documentation translation project](https://translate.openstack.org/version-group/view/doc-resources) (not interface). 

> @bkeepers Does it work well with vanilla Jekyll on GitHub Pages?
Jekyll parses markdown to render html, so translation work is done in markdown. Transifex can consume markdown files and [each paragraph gets its own entry](https://meta.discourse.org/t/maintaining-huge-translated-strings-in-transifex/27413) then translators can start working on their web app. After translation done, the files are available for pull. So technically there is no incompatibility. Here is [how it works](https://www.transifex.com/how-it-works/). For more automation, there is [Github integration](https://docs.transifex.com/integrations/github/).

Their main advantage is to help open source efficiently achieve [Crowdsourcing Translations](https://docs.transifex.com/guides/crowdsourcing-translations)

**Comment by bkeepers** [comment]
Thanks @joelhandwell for more info. GitHub's docs team is actually in the middle of evaluating systems for managing translations. I saw a demo of @yarnpkg's setup, which is using [Crowdin](https://www.crowdin.com/), and was impressed (cc @thejameskyle).

What's most important to me is that the system is compatible with an open source workflow. Has anyone contributed to other open source projects that have a good setup for translation?

**Comment by thejameskyle** [comment]
I wrote down all of the setup work involved with syncing between Crowdin and GitHub here: https://github.com/thejameskyle/crowdin-sync

**Comment by thejameskyle** [comment]
We've been successful in getting the community to translate the Yarn website https://yarnpkg.com/en/org/translations using Crowdin and the process outlined in [crowdin-sync](https://github.com/thejameskyle/crowdin-sync). This is the status in just a few weeks:

screen shot 2017-02-21 at 3 15 39 pm

**Comment by thejameskyle** [comment]
Oh and note that Crowdin chokes on Jekyll frontmatter right now (I've spoken to them and they will be adding support for it in the near future), in the meantime I created [this script](https://github.com/yarnpkg/website/blob/master/scripts/normalize-translations.rb) to just copy the english version into each one.

**Comment by zeke** [comment]
It's great to see how quickly the community has translated yarn's docs. 🐈 

@thejameskyle did you create [crowdin-sync](https://github.com/thejameskyle/crowdin-sync) because [Crowdin's GitHub integration](https://support.crowdin.com/github-integration/) is lacking in some way?

**Comment by thejameskyle** [comment]
@zeke Oh cool, I think that must be new. I would like to try that out

**Comment by thejameskyle** [comment]
Crowdin just told me that Jekyll frontmatter support should be implemented next week btw

**Comment by Gotgot59** [comment]
I Can't find the link to make the French Translation. Someone can Help ? 
What is the tech chose to make it ?  crowdin-sync ?

**Comment by zeke** [comment]
@Gotgot59 this is still a work in progress, but if you'd like to get a head start, follow the instructions outlined in the new `translations.md` file: https://github.com/github/open-source-guide/blob/bf259a187a8a5a84a784d9fafc3221303173e168/docs/translations.md

**Comment by nayafia** [comment]
@bkeepers I added one more todo in your original post: document the process for future translations :wink:

File: docs/templates/translation-issue-template.md [hunk 1]
```diff
@@ -1,8 +1,10 @@
 Repository: https://github/[yourname]/open-source-guide
+Preview: https://[yourname].github.io/open-source-guide/
 Locale: `xx-YY`
 
 On the `gh-pages` branch of your fork:
 
+- [ ] Merge the `i18n` branch into `gh-pages` and push the branch to your fork. This should build the website and it should be visible at the preview url above.
```
Comment (webknjaz) [inline docs/templates/translation-issue-template.md]: I wish this work. But it doesn't work for me currently. Any ideas?
@bkeepers

Comment (bkeepers) [inline docs/templates/translation-issue-template.md]: Hey @webknjaz, sorry for the issue. It's been fixed in #348. If you merge the latest changes from the `i18n` branch into your gh-pages branch, it should work now.

**Comment by jhabdas** [comment]
Hello! Looking forward to digging into this code. But the following statements give me pause in doing so:

> There is a Jekyll plugin for multiple languages, but it's not supported on GitHub Pages. So we'll have to manually add the facilities for supporting translations.

It seems it would be easier to maintain and more beneficial for the community if one of the following were to occur:

- Work with core team to bake language support into Jekyll
- Work with plugin maintainers and GH to introduce a new whitelisted i18n plugin

As it stands the number of years of Jekyll development have introduced a lot of individual effort and varying approaches having someone champion this effort on the Jekyll side seems like the best course of action to help save time for others going forward.

Nevertheless, looking forward to seeing this go live. See also: https://developmentseed.org/blog/multilingual-jekyll-sites/

**Comment by bkeepers** [comment]
I've merged in all the latest changes and I think this branch is ready to merge. There are still some things to work out with the translation process, but we'll work through those in subsequent PRs and continue to refine the process as we start to promote official translations.

Note: Build is failing because of transient failures on external links. Ignoring for now

**Comment by d-damien** [comment]
Is it possible to define a fallback list of languages ? I suggest to make it similar to the $LANGUAGE var in bash. See `man 7 locale`.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -26,7 +26,7 @@ There are also many ways you can directly contribute to the guides (in descendin
 * Fix editorial inconsistencies or inaccuracies
 * Add stories, examples, or anecdotes that help illustrate a point
 * Revise language to be more approachable and friendly
-* Translate guides into other languages
```
Comment (ahmed6192) [inline CONTRIBUTING.md:29]: 1

**Pull request merged**